iocore/net/quic/QUICPacketProtectionKeyInfo.cc

```diff
@@ -0,0 +1,348 @@
+    return this->_is_server_key_available[index];
+  }
+}
+
+void
+QUICPacketProtectionKeyInfo::set_decryption_key_available(QUICKeyPhase phase)
+{
+  int index = static_cast<int>(phase);
+  if (this->_ctx == Context::SERVER) {
+    this->_is_client_key_available[index] = true;
+  } else {
+    this->_is_server_key_available[index] = true;
+    ;
+  }
+}
+
+const uint8_t *
+QUICPacketProtectionKeyInfo::decryption_key(QUICKeyPhase phase) const
+{
+  int index = static_cast<int>(phase);
+  if (this->_ctx == Context::SERVER) {
+    return this->_client_key[index];
+  } else {
+    return this->_server_key[index];
+  }
+}
+
+uint8_t *
+QUICPacketProtectionKeyInfo::decryption_key(QUICKeyPhase phase)
+{
+  return const_cast<uint8_t *>(const_cast<const QUICPacketProtectionKeyInfo *>(this)->decryption_key(phase));
+}
+
+size_t
+QUICPacketProtectionKeyInfo::decryption_key_len(QUICKeyPhase phase) const
+{
+  const QUIC_EVP_CIPHER *cipher;
+
+  switch (phase) {
+  case QUICKeyPhase::INITIAL:
+    cipher = this->_cipher_initial;
+    break;
+  default:
+    cipher = this->_cipher;
+    break;
+  }
+
+  return EVP_CIPHER_key_length(cipher);
+}
+
+const uint8_t *
+QUICPacketProtectionKeyInfo::decryption_iv(QUICKeyPhase phase) const
+{
+  int index = static_cast<int>(phase);
+  if (this->_ctx == Context::SERVER) {
+    return this->_client_iv[index];
+  } else {
+    return this->_server_iv[index];
+  }
+}
+
+uint8_t *
+QUICPacketProtectionKeyInfo::decryption_iv(QUICKeyPhase phase)
+{
+  return const_cast<uint8_t *>(const_cast<const QUICPacketProtectionKeyInfo *>(this)->decryption_iv(phase));
+}
+
+const size_t *
+QUICPacketProtectionKeyInfo::decryption_iv_len(QUICKeyPhase phase) const
+{
+  int index = static_cast<int>(phase);
+  if (this->_ctx == Context::SERVER) {
+    return &this->_client_iv_len[index];
+  } else {
+    return &this->_server_iv_len[index];
+  }
+}
+
+size_t *
+QUICPacketProtectionKeyInfo::decryption_iv_len(QUICKeyPhase phase)
+{
+  return const_cast<size_t *>(const_cast<const QUICPacketProtectionKeyInfo *>(this)->decryption_iv_len(phase));
+}
+
+void
+QUICPacketProtectionKeyInfo::set_cipher_initial(const QUIC_EVP_CIPHER *cipher)
+{
+  this->_cipher_initial = cipher;
+}
+
+void
+QUICPacketProtectionKeyInfo::set_cipher(const QUIC_EVP_CIPHER *cipher, size_t tag_len)
+{
+  this->_cipher  = cipher;
+  this->_tag_len = tag_len;
+}
+
+const QUIC_EVP_CIPHER *
+QUICPacketProtectionKeyInfo::get_cipher_for_hp(QUICKeyPhase phase) const
+{
+  switch (phase) {
+  case QUICKeyPhase::INITIAL:
+    return this->_cipher_for_hp_initial;
+  default:
+    return this->_cipher_for_hp;
+  }
+}
+
+void
+QUICPacketProtectionKeyInfo::set_cipher_for_hp_initial(const QUIC_EVP_CIPHER *cipher)
+{
+  this->_cipher_for_hp_initial = cipher;
+}
+
+void
+QUICPacketProtectionKeyInfo::set_cipher_for_hp(const QUIC_EVP_CIPHER *cipher)
+{
+  this->_cipher_for_hp = cipher;
+}
+
+const uint8_t *
+QUICPacketProtectionKeyInfo::encryption_key_for_hp(QUICKeyPhase phase) const
```

Comment:
```
uint8_t *
QUICPacketProtectionKeyInfo::encryption_key_for_hp(QUICKeyPhase phase)

const uint8_t *
QUICPacketProtectionKeyInfo::encryption_key_for_hp(QUICKeyPhase phase) const
```

Not sure do we need these two function.

Reply:
I wanted to make sure only classes that have the non-const object can update the information. By having the two, key *users* cannot break the information.